In [30]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [31]:
df = pd.read_csv("E:/postgre/CarPrice_Assignment.csv", header = 0)

In [32]:
df

car_ID  symboling                   CarName fueltype aspiration  \
0         1          3        alfa-romero giulia      gas        std   
1         2          3       alfa-romero stelvio      gas        std   
2         3          1  alfa-romero Quadrifoglio      gas        std   
3         4          2               audi 100 ls      gas        std   
4         5          2                audi 100ls      gas        std   
..      ...        ...                       ...      ...        ...   
200     201         -1           volvo 145e (sw)      gas        std   
201     202         -1               volvo 144ea      gas      turbo   
202     203         -1               volvo 244dl      gas        std   
203     204         -1                 volvo 246   diesel      turbo   
204     205         -1               volvo 264gl      gas      turbo   

    doornumber      carbody drivewheel enginelocation  wheelbase  ...  \
0          two  convertible        rwd          front       88.6  ...   
1          two  convertible        rwd          front       88.6  ...   
2          two    hatchback        rwd          front       94.5  ...   
3         four        sedan        fwd          front       99.8  ...   
4         four        sedan        4wd          front       99.4  ...   
..         ...          ...        ...            ...        ...  ...   
200       four        sedan        rwd          front      109.1  ...   
201       four        sedan        rwd          front      109.1  ...   
202       four        sedan        rwd          front      109.1  ...   
203       four        sedan        rwd          front      109.1  ...   
204       four        sedan        rwd          front      109.1  ...   

     enginesize  fuelsystem  boreratio  stroke compressionratio horsepower  \
0           130        mpfi       3.47    2.68              9.0        111   
1           130        mpfi       3.47    2.68              9.0        111   
2           152        mpfi       2.68    3.47              9.0        154   
3           109        mpfi       3.19    3.40             10.0        102   
4           136        mpfi       3.19    3.40              8.0        115   
..          ...         ...        ...     ...              ...        ...   
200         141        mpfi       3.78    3.15              9.5        114   
201         141        mpfi       3.78    3.15              8.7        160   
202         173        mpfi       3.58    2.87              8.8        134   
203         145         idi       3.01    3.40             23.0        106   
204         141        mpfi       3.78    3.15              9.5        114   

     peakrpm citympg  highwaympg    price  
0       5000      21          27  13495.0  
1       5000      21          27  16500.0  
2       5000      19          26  16500.0  
3       5500      24          30  13950.0  
4       5500      18          22  17450.0  
..       ...     ...         ...      ...  
200     5400      23          28  16845.0  
201     5300      19          25  19045.0  
202     5500      18          23  21485.0  
203     4800      26          27  22470.0  
204     5400      19          25  22625.0  

[205 rows x 26 columns]

In [33]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 205 entries, 0 to 204
Data columns (total 26 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   car_ID            205 non-null    int64  
 1   symboling         205 non-null    int64  
 2   CarName           205 non-null    object 
 3   fueltype          205 non-null    object 
 4   aspiration        205 non-null    object 
 5   doornumber        205 non-null    object 
 6   carbody           205 non-null    object 
 7   drivewheel        205 non-null    object 
 8   enginelocation    205 non-null    object 
 9   wheelbase         205 non-null    float64
 10  carlength         205 non-null    float64
 11  carwidth          205 non-null    float64
 12  carheight         205 non-null    float64
 13  curbweight        205 non-null    int64  
 14  enginetype        205 non-null    object 
 15  cylindernumber    205 non-null    object 
 16  enginesize        205 non-null    int64  
 1

As we can see there are no null values in this dataset. 

In [34]:
df.corr()

C:\Users\dell\AppData\Local\Temp\ipykernel_18792\1134722465.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


car_ID  symboling  wheelbase  carlength  carwidth  \
car_ID            1.000000  -0.151621   0.129729   0.170636  0.052387   
symboling        -0.151621   1.000000  -0.531954  -0.357612 -0.232919   
wheelbase         0.129729  -0.531954   1.000000   0.874587  0.795144   
carlength         0.170636  -0.357612   0.874587   1.000000  0.841118   
carwidth          0.052387  -0.232919   0.795144   0.841118  1.000000   
carheight         0.255960  -0.541038   0.589435   0.491029  0.279210   
curbweight        0.071962  -0.227691   0.776386   0.877728  0.867032   
enginesize       -0.033930  -0.105790   0.569329   0.683360  0.735433   
boreratio         0.260064  -0.130051   0.488750   0.606454  0.559150   
stroke           -0.160824  -0.008735   0.160959   0.129533  0.182942   
compressionratio  0.150276  -0.178515   0.249786   0.158414  0.181129   
horsepower       -0.015006   0.070873   0.353294   0.552623  0.640732   
peakrpm          -0.203789   0.273606  -0.360469  -0.287242 -0.220012   
citympg           0.015940  -0.035823  -0.470414  -0.670909 -0.642704   
highwaympg        0.011255   0.034606  -0.544082  -0.704662 -0.677218   
price            -0.109093  -0.079978   0.577816   0.682920  0.759325   

                  carheight  curbweight  enginesize  boreratio    stroke  \
car_ID             0.255960    0.071962   -0.033930   0.260064 -0.160824   
symboling         -0.541038   -0.227691   -0.105790  -0.130051 -0.008735   
wheelbase          0.589435    0.776386    0.569329   0.488750  0.160959   
carlength          0.491029    0.877728    0.683360   0.606454  0.129533   
carwidth           0.279210    0.867032    0.735433   0.559150  0.182942   
carheight          1.000000    0.295572    0.067149   0.171071 -0.055307   
curbweight         0.295572    1.000000    0.850594   0.648480  0.168790   
enginesize         0.067149    0.850594    1.000000   0.583774  0.203129   
boreratio          0.171071    0.648480    0.583774   1.000000 -0.055909   
stroke            -0.055307    0.168790    0.203129  -0.055909  1.000000   
compressionratio   0.261214    0.151362    0.028971   0.005197  0.186110   
horsepower        -0.108802    0.750739    0.809769   0.573677  0.080940   
peakrpm           -0.320411   -0.266243   -0.244660  -0.254976 -0.067964   
citympg           -0.048640   -0.757414   -0.653658  -0.584532 -0.042145   
highwaympg        -0.107358   -0.797465   -0.677470  -0.587012 -0.043931   
price              0.119336    0.835305    0.874145   0.553173  0.079443   

                  compressionratio  horsepower   peakrpm   citympg  \
car_ID                    0.150276   -0.015006 -0.203789  0.015940   
symboling                -0.178515    0.070873  0.273606 -0.035823   
wheelbase                 0.249786    0.353294 -0.360469 -0.470414   
carlength                 0.158414    0.552623 -0.287242 -0.670909   
carwidth                  0.181129    0.640732 -0.220012 -0.642704   
carheight                 0.261214   -0.108802 -0.320411 -0.048640   
curbweight                0.151362    0.750739 -0.266243 -0.757414   
enginesize                0.028971    0.809769 -0.244660 -0.653658   
boreratio                 0.005197    0.573677 -0.254976 -0.584532   
stroke                    0.186110    0.080940 -0.067964 -0.042145   
compressionratio          1.000000   -0.204326 -0.435741  0.324701   
horsepower               -0.204326    1.000000  0.131073 -0.801456   
peakrpm                  -0.435741    0.131073  1.000000 -0.113544   
citympg                   0.324701   -0.801456 -0.113544  1.000000   
highwaympg                0.265201   -0.770544 -0.054275  0.971337   
price                     0.067984    0.808139 -0.085267 -0.685751   

                  highwaympg     price  
car_ID              0.011255 -0.109093  
symboling           0.034606 -0.079978  
wheelbase          -0.544082  0.577816  
carlength          -0.704662  0.682920  
carwidth           -0.677218  0.759325  
carheight          -0.107358  0.119336  
curbwei

In [35]:
df.describe()

car_ID   symboling   wheelbase   carlength    carwidth   carheight  \
count  205.000000  205.000000  205.000000  205.000000  205.000000  205.000000   
mean   103.000000    0.834146   98.756585  174.049268   65.907805   53.724878   
std     59.322565    1.245307    6.021776   12.337289    2.145204    2.443522   
min      1.000000   -2.000000   86.600000  141.100000   60.300000   47.800000   
25%     52.000000    0.000000   94.500000  166.300000   64.100000   52.000000   
50%    103.000000    1.000000   97.000000  173.200000   65.500000   54.100000   
75%    154.000000    2.000000  102.400000  183.100000   66.900000   55.500000   
max    205.000000    3.000000  120.900000  208.100000   72.300000   59.800000   

        curbweight  enginesize   boreratio      stroke  compressionratio  \
count   205.000000  205.000000  205.000000  205.000000        205.000000   
mean   2555.565854  126.907317    3.329756    3.255415         10.142537   
std     520.680204   41.642693    0.270844    0.313597          3.972040   
min    1488.000000   61.000000    2.540000    2.070000          7.000000   
25%    2145.000000   97.000000    3.150000    3.110000          8.600000   
50%    2414.000000  120.000000    3.310000    3.290000          9.000000   
75%    2935.000000  141.000000    3.580000    3.410000          9.400000   
max    4066.000000  326.000000    3.940000    4.170000         23.000000   

       horsepower      peakrpm     citympg  highwaympg         price  
count  205.000000   205.000000  205.000000  205.000000    205.000000  
mean   104.117073  5125.121951   25.219512   30.751220  13276.710571  
std     39.544167   476.985643    6.542142    6.886443   7988.852332  
min     48.000000  4150.000000   13.000000   16.000000   5118.000000  
25%     70.000000  4800.000000   19.000000   25.000000   7788.000000  
50%     95.000000  5200.000000   24.000000   30.000000  10295.000000  
75%    116.000000  5500.000000   30.000000   34.000000  16503.000000  
max    288.000000  6600.000000   49.000000   54.000000  45400.000000

Generating dummy values

In [36]:
df = pd.get_dummies(df, columns = ['fueltype','aspiration','doornumber','carbody','drivewheel','enginelocation','fuelsystem','enginetype','cylindernumber'], drop_first = True)

In [37]:
df

car_ID  symboling                   CarName  wheelbase  carlength  \
0         1          3        alfa-romero giulia       88.6      168.8   
1         2          3       alfa-romero stelvio       88.6      168.8   
2         3          1  alfa-romero Quadrifoglio       94.5      171.2   
3         4          2               audi 100 ls       99.8      176.6   
4         5          2                audi 100ls       99.4      176.6   
..      ...        ...                       ...        ...        ...   
200     201         -1           volvo 145e (sw)      109.1      188.8   
201     202         -1               volvo 144ea      109.1      188.8   
202     203         -1               volvo 244dl      109.1      188.8   
203     204         -1                 volvo 246      109.1      188.8   
204     205         -1               volvo 264gl      109.1      188.8   

     carwidth  carheight  curbweight  enginesize  boreratio  ...  \
0        64.1       48.8        2548         130       3.47  ...   
1        64.1       48.8        2548         130       3.47  ...   
2        65.5       52.4        2823         152       2.68  ...   
3        66.2       54.3        2337         109       3.19  ...   
4        66.4       54.3        2824         136       3.19  ...   
..        ...        ...         ...         ...        ...  ...   
200      68.9       55.5        2952         141       3.78  ...   
201      68.8       55.5        3049         141       3.78  ...   
202      68.9       55.5        3012         173       3.58  ...   
203      68.9       55.5        3217         145       3.01  ...   
204      68.9       55.5        3062         141       3.78  ...   

     enginetype_ohc  enginetype_ohcf  enginetype_ohcv  enginetype_rotor  \
0                 0                0                0                 0   
1                 0                0                0                 0   
2                 0                0                1                 0   
3                 1                0                0                 0   
4                 1                0                0                 0   
..              ...              ...              ...               ...   
200               1                0                0                 0   
201               1                0                0                 0   
202               0                0                1                 0   
203               1                0                0                 0   
204               1                0                0                 0   

     cylindernumber_five  cylindernumber_four  cylindernumber_six  \
0                      0                    1                   0   
1                      0                    1                   0   
2                      0                    0                   1   
3                      0                    1                   0   
4                      1                    0                   0   
..                   ...                  ...                 ...   
200                    0                    1                   0   
201                    0                    1                   0   
202                    0                    0                   1   
203                    0                    0                   1   
204                    0                    1                   0   

     cylindernumber_three  cylindernumber_twelve  cylindernumber_two  
0                       0                      0                   0  
1                       0                      0                   0  
2                       0                      0                   0  
3                       0                      0                   0  
4                       0                      0                   0  
..                    ...                    ...                 ...  
200                     0                      0                   0  
2

## Multiple Linear Regression

In [38]:
from sklearn.linear_model import LinearRegression

In [40]:
del df['CarName']

In [41]:
x_multi = df.drop("price", axis = 1)

In [42]:
y_multi = df['price']

In [43]:
from sklearn.model_selection import train_test_split

In [44]:
x_train, x_test, y_train, y_test = train_test_split(x_multi, y_multi, test_size = 0.2, random_state = 0)

In [45]:
print(x_train.shape, x_test.shape, y_train.shape, y_test.shape)

(164, 44) (41, 44) (164,) (41,)


In [46]:
lm_a = LinearRegression()

In [47]:
lm_a.fit(x_train, y_train)

LinearRegression()

In [48]:
y_test_a = lm_a.predict(x_test)

In [49]:
y_train_a = lm_a.predict(x_train)

In [50]:
from sklearn.metrics import r2_score

In [51]:
r2_score(y_test, y_test_a)

0.8781649838248071

In [52]:
r2_score(y_train, y_train_a)

0.9588893127277134

**Can be improved further. Let us use Ridge Regression and Lasso.**

In [53]:
from sklearn import preprocessing

In [54]:
scaler = preprocessing.StandardScaler().fit(x_train)

In [55]:
x_train_s = scaler.transform(x_train)

In [56]:
x_test_s = scaler.transform(x_test)

In [57]:
from sklearn.linear_model import Ridge

In [58]:
lm_r = Ridge(alpha = 0.5)
lm_r.fit(x_train_s, y_train)

Ridge(alpha=0.5)

In [59]:
r2_score(y_test, lm_r.predict(x_test_s))

0.8786868971821522

In [60]:
from sklearn.model_selection import validation_curve

In [61]:
param_range = np.logspace(-2,8,100)

In [62]:
train_scores, test_scores = validation_curve(
    estimator=Ridge(),
    X=x_train_s,
    y=y_train,
    param_name="alpha",
    param_range=param_range,
    scoring='r2'
)

In [63]:
print(train_scores)
print(test_scores)

[[9.63672637e-01 9.63302335e-01 9.64272655e-01 9.63224947e-01
  9.58882710e-01]
 [9.63672166e-01 9.63301477e-01 9.64271914e-01 9.63224788e-01
  9.58882548e-01]
 [9.63671434e-01 9.63300144e-01 9.64270765e-01 9.63224537e-01
  9.58882292e-01]
 [9.63670301e-01 9.63298090e-01 9.64268994e-01 9.63224140e-01
  9.58881893e-01]
 [9.63668560e-01 9.63294947e-01 9.64266286e-01 9.63223515e-01
  9.58881270e-01]
 [9.63665905e-01 9.63290184e-01 9.64262184e-01 9.63222533e-01
  9.58880303e-01]
 [9.63661898e-01 9.63283046e-01 9.64256041e-01 9.63220998e-01
  9.58878814e-01]
 [9.63655922e-01 9.63272492e-01 9.64246960e-01 9.63218607e-01
  9.58876533e-01]
 [9.63647132e-01 9.63257135e-01 9.64233738e-01 9.63214902e-01
  9.58873069e-01]
 [9.63634404e-01 9.63235190e-01 9.64214817e-01 9.63209198e-01
  9.58867854e-01]
 [9.63616296e-01 9.63204468e-01 9.64188244e-01 9.63200489e-01
  9.58860083e-01]
 [9.63591015e-01 9.63162411e-01 9.64151667e-01 9.63187314e-01
  9.58848628e-01]
 [9.63556400e-01 9.63106183e-01 9.641023

In [64]:
train_mean = np.mean(train_scores, axis = 1)

In [65]:
test_mean = np.mean(test_scores, axis = 1)

In [66]:
max(test_mean)

0.9053242709998456

In [67]:
max(train_mean)

0.9626710569878669

In [68]:
np.where(test_mean==max(test_mean))

(array([22], dtype=int64),)

In [77]:
lm_r_best = Ridge(alpha = param_range[22])

1.6681005372000592

In [85]:
lm_r_best = Ridge(alpha = param_range[22])

In [86]:
lm_r_best.fit(x_train_s, y_train)

Ridge(alpha=1.6681005372000592)

In [87]:
r2_score(y_test, lm_r_best.predict(x_test_s))

0.8737444350121568

It seems 87% accuracy is far as we can make it on this one. 